### ISSUE TITLE
BadSchemaError

### ISSUE THREAD

**Issue by zicameau** [taverntesting/tavern#18]
Okay so I've been trying to find information on this problem however it seems that your package is fairly new so I can't find much online.  Anyways, my problem is that I am trying to test my server however I keep ending up with badschemaerrors whenever I try to test my yaml file with pytest.  Here is my yaml file:

```
# test_sz.tavern.yaml

---

test_name: Testing Authentication Module

stages:
 request:
  url: https://localhost:8443/api/authenticate
  method: POST
  data: 
   password: *****
   username: *****
   thirdPartyLogin: false
   forceLogin: false
  headers:
   content-type: application/x-www-form-urlencoded
 response:
  status_code: 200

---
```

and here is what I get in response to running py.test in the same directory as the file:

```
C:\Users\TON\Documents\Rest API Pulling Documentation>py.test
============================= test session starts =============================
platform win32 -- Python 3.6.4, pytest-3.3.2, py-1.5.2, pluggy-0.6.0
rootdir: C:\Users\TON\Documents\Rest API Pulling Documentation, inifile:
plugins: tavern-0.2.1
collected 0 items / 1 errors

=================================== ERRORS ====================================
____________________ ERROR collecting test_sz.tavern.yaml _____________________
..\..\appdata\local\programs\python\python36-32\lib\site-packages\tavern\schemas\files.py:38: in verify_generic
    verifier.validate()
..\..\appdata\local\programs\python\python36-32\lib\site-packages\pykwalify\core.py:176: in validate
    error_msg=u'.\n - '.join(self.validation_errors)))
E   pykwalify.errors.SchemaError: <SchemaError: error code 2: Schema validation failed:
E    - Value '{'request': {'data': {'forceLogin': False, 'password': 'Panduit12', 'thirdPartyLogin': False, 'username': 'admin'}, 'headers': {'content-type': 'application/x-www-form-urlencoded'}, 'method': 'POST', 'url': 'https://localhost:8443/api/authenticate'}, 'response': {'status_code': 200}}' is not a list. Value path: '/stages'.: Path: '/'>

The above exception was the direct cause of the following exception:
..\..\appdata\local\programs\python\python36-32\lib\site-packages\tavern\testutils\pytesthook.py:49: in collect
    verify_tests(test_spec)
..\..\appdata\local\programs\python\python36-32\lib\site-packages\tavern\schemas\files.py:82: in verify_tests
    verify_generic(test_tmp, schema_filename)
..\..\appdata\local\programs\python\python36-32\lib\site-packages\tavern\schemas\files.py:41: in verify_generic
    raise_from(BadSchemaError(), e)
..\..\appdata\local\programs\python\python36-32\lib\site-packages\future\utils\__init__.py:398: in raise_from
    exec(execstr, myglobals, mylocals)
<string>:1: in <module>
    ???
E   tavern.util.exceptions.BadSchemaError
------------------------------- Captured stderr -------------------------------
validation.invalid
 --- All found errors ---
["Value '{'request': {'data': {'forceLogin': False, 'password': '****', 'thirdPartyLogin': False, 'username': '****'}, 'headers': {'content-type': 'application/x-www-form-urlencoded'}, 'method': 'POST', 'url': 'https://localhost:8443/api/authenticate'}, 'response': {'status_code': 200}}' is not a list. Value path: '/stages'"]
Error validating C:\Users\TON\AppData\Local\Temp\tmp8irts1la.yaml
Traceback (most recent call last):
  File "c:\users\ton\appdata\local\programs\python\python36-32\lib\site-packages\tavern\schemas\files.py", line 38, in verify_generic
    verifier.validate()
  File "c:\users\ton\appdata\local\programs\python\python36-32\lib\site-packages\pykwalify\core.py", line 176, in validate
    error_msg=u'.\n - '.join(self.validation_errors)))
pykwalify.errors.SchemaError: <SchemaError: error code 2: Schema validation failed:
 - Value '{'request': {'data': {'forceLogin': False, 'password': '****', 'thirdPartyLogin': False, 'username': '****'}, 'headers': {'content-type': 'application/x-www-form-urlencoded'}, 'method': 'POST', 'url': 'https://localhost:8443/api/authenticate'}, 'response': {'status_code': 200}}' is not a list. Value path: '/stages'.: Path: '/'>
!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!
=========================== 1 error in 0.51 seconds ===========================
```

It keeps talking about how the value is not a list which makes no sense to me.  Thanks for taking the time to take a look at this and I look forward to your responses! :)

**Comment by benhowes** [comment]
@zicameau - I'm not a tavern expert, however it looks from the examples that your `stages` are not a list of stages, try something like:

```
stages:
  - name: Login step
    request:
      url: https://localhost:8443/api/authenticate
      method: POST
      data: 
       password: *****
       username: *****
       thirdPartyLogin: false
       forceLogin: false
      headers:
       content-type: application/x-www-form-urlencoded
    response:
      status_code: 200
```

The `-` indicates a list element in yaml syntax. Let me know how you get on!

**Comment by zicameau** [comment]
@benhowes 

This works if I call the file using the python code as shown in the documentation examples, but it doesn't work if I try to use py.test in the directory that the test file is located in:

```
C:\Users\TON\Documents\Rest API Pulling Documentation>py.test
============================= test session starts =============================
platform win32 -- Python 3.6.4, pytest-3.3.2, py-1.5.2, pluggy-0.6.0
rootdir: C:\Users\TON\Documents\Rest API Pulling Documentation, inifile:
plugins: tavern-0.2.1
collected 0 items / 1 errors

=================================== ERRORS ====================================
____________________ ERROR collecting test_sz.tavern.yaml _____________________
..\..\appdata\local\programs\python\python36-32\lib\site-packages\tavern\testutils\pytesthook.py:49: in collect
    verify_tests(test_spec)
..\..\appdata\local\programs\python\python36-32\lib\site-packages\tavern\schemas\files.py:82: in verify_tests
    verify_generic(test_tmp, schema_filename)
..\..\appdata\local\programs\python\python36-32\lib\site-packages\tavern\schemas\files.py:34: in verify_generic
    extensions=[extension_module_filename],
..\..\appdata\local\programs\python\python36-32\lib\site-packages\pykwalify\core.py:123: in __init__
    raise CoreError(u"No source file/data was loaded")
E   pykwalify.errors.CoreError: <CoreError: error code 3: No source file/data was loaded: Path: '/'>
!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!
=========================== 1 error in 0.49 seconds ===========================
```

I appreciate you're help by the way! Also, I saw on the website [https://taverntesting.github.io/documentation](here) that the only supported keys are url, json, params, data, headers, and method. I would like to be able to add the verify key to that somehow in order to be able to test a local server that doesn't have ssl verification encoded into it. 

I was wondering if there might be a work around to getting this to work with how the packages is already setup? Or if there is a way we might be able to add this key to the functionality of tavern itself and possibly add some other keyword arguments too like auth, file etc.

**Comment by michaelboulton** [comment]
Can you post the entire `test_sz.tavern.yaml` so I can see if I can figure out what's wrong?

As for the keys - it only supported those keys because they were the ones I was using. I've just added support for the 'verify' key in 0.2.2. I can add the 'auth' header relatively easily as well - do you just pass a tuple like `("Basic", "abc123")` to it? I tend to just explicitly use something like `Authorization: Basic abc123` in the `header` block.

The `files` is a bit more complicated because it takes file objects, it would take a fair amount of work to add loading the files and adding proper error handling etc. but it's something that could be added in future.

**Comment by zicameau** [comment]
Here is the total yaml file that I have made:

```
# test_sz.tavern.yaml

---
test_name: Testing Authentication Module

stages:
- name: Testing
  request:
    url: https://localhost:8443/api/authenticate
    method: POST
    data: 
      password: *****
      username: *****
      thirdPartyLogin: false
      forceLogin: false
    headers:
      content-type: application/x-www-form-urlencoded
    verify: false
  response:
    status_code: 200

---
```

And here is the response that I get from the command line when calling py.test in the same directory as the yaml file:

```
C:\Users\TON\Documents\Rest API Pulling Documentation>py.test
============================= test session starts =============================
platform win32 -- Python 3.6.4, pytest-3.3.2, py-1.5.2, pluggy-0.6.0
rootdir: C:\Users\TON\Documents\Rest API Pulling Documentation, inifile:
plugins: tavern-0.2.3
collected 0 items / 1 errors

=================================== ERRORS ====================================
____________________ ERROR collecting test_sz.tavern.yaml _____________________
..\..\appdata\local\programs\python\python36-32\lib\site-packages\tavern\testutils\pytesthook.py:51: in collect
    verify_tests(test_spec)
..\..\appdata\local\programs\python\python36-32\lib\site-packages\tavern\schemas\files.py:82: in verify_tests
    verify_generic(test_tmp, schema_filename)
..\..\appdata\local\programs\python\python36-32\lib\site-packages\tavern\schemas\files.py:34: in verify_generic
    extensions=[extension_module_filename],
..\..\appdata\local\programs\python\python36-32\lib\site-packages\pykwalify\core.py:123: in __init__
    raise CoreError(u"No source file/data was loaded")
E   pykwalify.errors.CoreError: <CoreError: error code 3: No source file/data was loaded: Path: '/'>
!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!
=========================== 1 error in 0.44 seconds ===========================
```

Thanks for adding the verify keyword, that was my main request really.   The auth I feel like should have it's own block for username and password however the explicit way that you have done it could work too.  Also the files keyword is unimportant for me as of now, I was just curious to see what would be needed in order to implement that keyword.

**Comment by michaelboulton** [comment]
I think that's just because you have an extra `---` at the bottom of the file - that starts a new yaml document but because there's nothing in it then pykwalify complains.

It shouldn't raise an error though, that needs fixing

**Comment by michaelboulton** [comment]
That issue should now be fixed in 0.2.5
